packages/experiment-tag/src/experiment.ts

```diff
@@ -8,443 +8,675 @@ import {
+      this.setUrlChangeListener([
+        ...this.localFlagKeys,
+        ...this.remoteFlagKeys,
+      ]);
+    }
 
-        // Apply anti-flicker CSS to prevent UI flicker
-        applyAntiFlickerCss();
-      }
-    } else {
-      // Add locally evaluable flags to the local flag set
-      localFlagKeys.add(key);
-    }
-
-    flag.metadata = metadata;
-  });
-
-  initialFlags = JSON.stringify(parsedFlags);
-
-  // initialize the experiment
-  globalScope.webExperiment = Experiment.initialize(apiKey, {
-    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
-    // @ts-ignore
-    internalInstanceNameSuffix: 'web',
-    initialFlags: initialFlags,
-    // timeout for fetching remote flags
-    fetchTimeoutMillis: 1000,
-    pollOnStart: false,
-    fetchOnStart: false,
-    ...config,
-  });
-
-  // If no integration has been set, use an amplitude integration.
-  if (!globalScope.experimentIntegration) {
-    const connector = AnalyticsConnector.getInstance('$default_instance');
-    globalScope.experimentIntegration = new AmplitudeIntegrationPlugin(
-      apiKey,
-      connector,
-      0,
-    );
-  }
-  globalScope.experimentIntegration.type = 'integration';
-  globalScope.webExperiment.addPlugin(globalScope.experimentIntegration);
-  globalScope.webExperiment.setUser(user);
+    // apply local variants
+    this.applyVariants(this.localFlagKeys);
 
-  setUrlChangeListener(new Set([...localFlagKeys, ...remoteFlagKeys]));
+    if (
+      !this.isRemoteBlocking ||
+      !this.config.applyAntiFlickerForRemoteBlocking
+    ) {
+      // Remove anti-flicker css if remote flags are not blocking
+      this.globalScope?.document.getElementById?.('amp-exp-css')?.remove();
+    }
 
-  // apply local variants
-  applyVariants(globalScope.webExperiment.all(), localFlagKeys);
+    if (this.remoteFlagKeys.length === 0) {
+      return;
+    }
 
-  if (!isRemoteBlocking) {
-    // Remove anti-flicker css if remote flags are not blocking
-    globalScope.document.getElementById?.('amp-exp-css')?.remove();
+    await this.fetchRemoteFlags();
+    // apply remote variants - if fetch is unsuccessful, fallback order: 1. localStorage flags, 2. initial flags
+    this.applyVariants(this.remoteFlagKeys);
   }
 
-  if (remoteFlagKeys.size === 0) {
-    return;
+  /**
+   * Get the underlying ExperimentClient instance.
+   */
+  public getExperimentClient(): ExperimentClient | undefined {
+    return this.experimentClient;
   }
 
-  try {
-    await globalScope.webExperiment.doFlags();
-  } catch (error) {
-    console.warn('Error fetching remote flags:', error);
+  /**
+   * Set the context for evaluating experiments.
+   * If user is not provided, the current user is used.
+   * If currentUrl is not provided, the current URL is used.
+   * @param user
+   * @param currentUrl
+   */
+  public setContext(
+    user: ExperimentUser | undefined = undefined,
+    currentUrl: string | undefined = undefined,
```

Comment:
We should make this an object `WebExperimentContext` or something where each of these are optional fields.